Core/StitchCoreSDK/Sources/StitchCoreSDK/Auth/Internal/CoreStitchAuth+LoginAndLink.swift

```diff
@@ -0,0 +1,280 @@
+            // but they are logged in with their older credentials.
+            if asLinkRequest {
+                self.activeUserAuthInfo = nil
+                self.activeUser = nil
+
+                if let authInfo = oldAuthInfo {
+                    self.activeUserAuthInfo = StoreAuthInfo.init(
+                        withAuthInfo: authInfo,
+                        withProviderType: type(of: credential).providerType,
+                        withProviderName: credential.providerName).toAuthInfo
+
+                    self.activeUser = self.userFactory.makeUser(
+                        withID: authInfo.userID,
+                        withLoggedInProviderType: type(of: credential).providerType,
+                        withLoggedInProviderName: credential.providerName,
+                        withUserProfile: authInfo.userProfile,
+                        withIsLoggedIn: authInfo.isLoggedIn,
+                        withLastAuthActivity: authInfo.lastAuthActivity ?? 0.0)
+                }
+            } else if !self.loggedInUsersAuthInfo.isEmpty {
+                self.activeUserAuthInfo = oldAuthInfo
+                self.activeUser = oldUser
+            } else {
+                clearAllAuth()
+            }
+
+            throw err
+        }
+
+        // Update the lastAuthActivity of the old active user
+        if let oldAuthInfo = oldAuthInfo {
+            if let oldIndex = loggedInUsersAuthInfo.firstIndex(where: {$0.userID == oldAuthInfo.userID}) {
+                let oldAuthInfo = StoreAuthInfo.init(withAuthInfo: oldAuthInfo, withNewTime: true).toAuthInfo
+                loggedInUsersAuthInfo[oldIndex] = oldAuthInfo
+            }
+        }
+
+        // Finally set the info and user
+        var newAuthInfo = StoreAuthInfo.init(
+            withAPIAuthInfo: newAPIAuthInfo,
+            withExtendedAuthInfo: ExtendedAuthInfoImpl.init(loggedInProviderType: type(of: credential).providerType,
+                                                            loggedInProviderName: credential.providerName,
+                                                            userProfile: profile)).toAuthInfo
+        newAuthInfo = StoreAuthInfo.init(withAuthInfo: newAuthInfo, withNewTime: true).toAuthInfo
+
+        let newUser = self.userFactory.makeUser(withID: newAuthInfo.userID,
+                                                withLoggedInProviderType: newAuthInfo.loggedInProviderType,
+                                                withLoggedInProviderName: newAuthInfo.loggedInProviderName,
+                                                withUserProfile: newAuthInfo.userProfile,
+                                                withIsLoggedIn: newAuthInfo.isLoggedIn,
+                                                withLastAuthActivity: newAuthInfo.lastAuthActivity ?? 0.0)
+
+        // Persist auth info to storage, and return the resulting StitchUser
+        do {
```

Comment:
[optional code cleanup] I think you could refactor this whole part of the code to be a single try-catch block, there's some duplicated code in the `catch` blocks and they're essentially doing the same thing. See JS and Java